README.md
二进制与字符串互转。

因为 js 的字符编码其实是一个 UTF16String，不是 utf8。所以在 js 内部自己转换，与传输到其他地方的转换是不一样的（因为传输到其他地方一般是 utf8 编码的 base64）。

可以获取到三种状态

1. 原始字符串
2. 中间状态的 arrayBuffer
3. 转换后可以直接传输的 base64

api 文档：https://sklme.github.io/binary-string/

# 安装

```shell
npm install @iskl/binary-string
yarn add @iskl/binary-string
pnpm add @iskl/binary-string
```

# 使用方法

## 实例化

### uf8string

将字符串作为 utf8 编码转换

```typescript
import { UTF8Transformer } from '@iskl/binary-string';

const utf8Transformer = new UTF8Transformer('一二三123abc⑴⑵⑶☆★✓');
// 传入数据类型
const base64SourceDomTransformer = new UTF8Transformer(
  '5LiA5LqM5LiJMTIzYWJj4pG04pG14pG24piG4piF4pyT',
  'base64String',
);

utf8Transformer.encode(); // 5LiA5LqM5LiJMTIzYWJj4pG04pG14pG24piG4piF4pyT
utf8Transformer.decode(); // 一二三123abc⑴⑵⑶☆★✓
utf8Transformer.getArrayBuffer();

base64SourceDomTransformer.encode(); // 5LiA5LqM5LiJMTIzYWJj4pG04pG14pG24piG4piF4pyT
base64SourceDomTransformer.decode(); // 一二三123abc⑴⑵⑶☆★✓
base64SourceDomTransformer.getArrayBuffer();
```

### dom-string

将字符串作为 UTF16 编码转换

```typescript
import { DomStringTransformer } from '@iskl/binary-string';

const domTransformer = new DomStringTransformer('一二三123abc⑴⑵⑶☆★✓');
// 传入数据类型
const base64SourceDomTransformer = new DomStringTransformer(
  'AE6MTglOMQAyADMAYQBiAGMAdCR1JHYkBiYFJhMn',
  'base64String',
);

domTransformer.encode(); // AE6MTglOMQAyADMAYQBiAGMAdCR1JHYkBiYFJhMn
domTransformer.decode(); // 一二三123abc⑴⑵⑶☆★✓
domTransformer.getArrayBuffer();

base64SourceDomTransformer.encode(); // AE6MTglOMQAyADMAYQBiAGMAdCR1JHYkBiYFJhMn
base64SourceDomTransformer.decode(); // 一二三123abc⑴⑵⑶☆★✓
base64SourceDomTransformer.getArrayBuffer();
```

## 简单使用

### uf8string

将字符串作为 utf8 编码转换

```typescript
import {
  stringToArrayBuffer, // 原始字符串转为中间二进制
  arrayBufferToBase64, // 中间二进制转为base64字符串
  base64ToArrayBuffer, // base64字符串转为中间二进制
  arrayBufferToString, // 中间二进制转为原始字符串
  encode, // 原始字符串转为base64字符串
  decode, // base64字符串转为原始字符串
} from '@iskl/binary-string/utf8-string';

encode('一二三123abc⑴⑵⑶☆★✓'); // < 5LiA5LqM5LiJMTIzYWJj4pG04pG14pG24piG4piF4pyT
decode('5LiA5LqM5LiJMTIzYWJj4pG04pG14pG24piG4piF4pyT'); // 一二三123abc⑴⑵⑶☆★✓
```

### dom-string

将字符串作为 UTF16 编码转换

```typescript
import {
  stringToArrayBuffer, // 原始字符串转为中间二进制
  arrayBufferToBase64, // 中间二进制转为base64字符串
  base64ToArrayBuffer, // base64字符串转为中间二进制
  arrayBufferToString, // 中间二进制转为原始字符串
  encode, // 原始字符串转为base64字符串
  decode, // base64字符串转为原始字符串
} from '@iskl/binary-string/dom-string';

encode('一二三123abc⑴⑵⑶☆★✓'); // < AE6MTglOMQAyADMAYQBiAGMAdCR1JHYkBiYFJhMn
decode('AE6MTglOMQAyADMAYQBiAGMAdCR1JHYkBiYFJhMn'); // 一二三123abc⑴⑵⑶☆★✓
```


CHANGELOG.md
# Changelog

## [1.3.0](https://github.com/sklme/binary-string/compare/v1.2.0...v1.3.0) (2023-02-05)


### Features

* remove postinstall ([ac10612](https://github.com/sklme/binary-string/commit/ac10612b9fce1f38da72b2a4aba1233da5a8d753))

## [1.2.0](https://github.com/sklme/binary-string/compare/v1.1.0...v1.2.0) (2023-02-05)


### Features

* relase as 1.2 ([ae2046d](https://github.com/sklme/binary-string/commit/ae2046d1c98f5b2146d9898fed96cb9165f6df85))

## [1.1.0](https://github.com/sklme/binary-string/compare/v1.0.0...v1.1.0) (2023-02-05)

### Features

- baseTransfomer 继承 ([d58c622](https://github.com/sklme/binary-string/commit/d58c622a81f122d689c2077de499e2cafaf2dad9))
- doc ([d348670](https://github.com/sklme/binary-string/commit/d3486705b8b362a609514bfb12929a12e0ee392e))
- **dom string:** encoder ([4047e16](https://github.com/sklme/binary-string/commit/4047e163f64324480ac03cfac3676c2d573b6c18))
- **utf8 string:** decode ([318c395](https://github.com/sklme/binary-string/commit/318c395ef26614b2c4b7f17a790bb375c56c49e2))
- utf8 算法 ([9a07dfa](https://github.com/sklme/binary-string/commit/9a07dfa5aaf624b3b0675c418cba9124ef17b1e9))
- 增加 typesVersions ，适配旧版本的 typescript ([a6f34d7](https://github.com/sklme/binary-string/commit/a6f34d701923c81f4e39fc8b32666d29c6470a3e))

### Bug Fixes

- build doc script ([bc01346](https://github.com/sklme/binary-string/commit/bc01346553483a6f186cc1d2860264e1aa340b05))

## 1.0.0 (2023-02-02)

### Features

- dom sting 的处理 ([4f5c81c](https://github.com/sklme/binary-string/commit/4f5c81c00e8a1bdba9ec5bebc1b74d27193fb431))
- **dom string:** string to arrayBuffer ([b5b0e0b](https://github.com/sklme/binary-string/commit/b5b0e0b65e016c50518bb1d0455208314e4fba63))


src/dom-string/index.spec.ts
import test, { ExecutionContext } from 'ava';
import {
  stringToArrayBuffer,
  arrayBufferToBase64,
  encode,
  base64ToArrayBuffer,
  arrayBufferToString,
  decode,
  Transformer,
} from '.';

const str = '一二三123abc⑴⑵⑶☆★✓';
const charCodeArray = [
  19968, 20108, 19977, 49, 50, 51, 97, 98, 99, 9332, 9333, 9334, 9734, 9733,
  10003,
];
const b64 = 'AE6MTglOMQAyADMAYQBiAGMAdCR1JHYkBiYFJhMn';

test('dom字符串转为arrayBuffer', (t) => {
  const buffer = stringToArrayBuffer(str);
  const u16 = new Uint16Array(buffer);

  t.deepEqual([...u16], charCodeArray);
});

test('arrayBuffer转为base64', (t) => {
  const { buffer } = Uint16Array.from(charCodeArray);

  const base64 = arrayBufferToBase64(buffer);

  t.is(b64, base64);
});

test('dom字符串encode', (t) => {
  t.is(b64, encode(str));
});

test('base64转为arrayBuffer', (t) => {
  const buffer = base64ToArrayBuffer(b64);

  t.deepEqual([...new Uint16Array(buffer)], charCodeArray);
});

test('arrayBufferToString', (t) => {
  const { buffer } = Uint16Array.from(charCodeArray);

  t.is(arrayBufferToString(buffer), str);
});

test('dom字符串decode', (t) => {
  t.is(decode(b64), str);
});

function testEncoder(encoder: Transformer, t: ExecutionContext) {
  t.is(encoder.decode(), str);
  t.is(encoder.encode(), b64);

  const arrayBuffer = encoder.getArrayBuffer();

  if (!arrayBuffer) {
    t.fail();
  } else {
    t.deepEqual([...new Uint16Array(arrayBuffer)], charCodeArray);
  }

  // return {
  //   encoded: encoder.encode(),
  //   decoded: encoder.decode(),
  //   arrayBuffer: encoder.getArrayBuffer(),
  // };
}

test('encoder: 来源为string', (t) => {
  const encoder = new Transformer(str);
  const passTypeEncoder = new Transformer(str, 'string');

  testEncoder(encoder, t);
  testEncoder(passTypeEncoder, t);
});

test('encoder: 来源为base64String', (t) => {
  const encoder = new Transformer(b64, 'base64String');
  testEncoder(encoder, t);
});

test('encoder: 来源为arrayBuffer', (t) => {
  const { buffer } = Uint16Array.from(charCodeArray);
  const encoder = new Transformer(buffer);
  const passTypeEncoder = new Transformer(buffer, 'arrayBuffer');

  testEncoder(encoder, t);
  testEncoder(passTypeEncoder, t);
});


src/utf8-string/index.spec.ts
import test, { ExecutionContext } from 'ava';
import {
  stringToArrayBuffer,
  arrayBufferToBase64,
  encode,
  base64ToArrayBuffer,
  arrayBufferToString,
  decode,
  Transformer,
} from '.';

const str = '一二三123abc⑴⑵⑶☆★✓';
const charCodeArray = [
  228, 184, 128, 228, 186, 140, 228, 184, 137, 49, 50, 51, 97, 98, 99, 226, 145,
  180, 226, 145, 181, 226, 145, 182, 226, 152, 134, 226, 152, 133, 226, 156,
  147,
];
const base64Str = '5LiA5LqM5LiJMTIzYWJj4pG04pG14pG24piG4piF4pyT';

test('stringToArrayBuffer', (t) => {
  const buffer = stringToArrayBuffer(str);

  const u8 = new Uint8Array(buffer);

  t.deepEqual([...u8], charCodeArray);
});

test('arrayBufferToBase64', (t) => {
  const b64 = arrayBufferToBase64(Uint8Array.from(charCodeArray).buffer);

  t.is(b64, base64Str);
});

test('encode', (t) => {
  const b64 = encode(str);

  t.is(b64, base64Str);
});

test('base64ToArrayBuffer', (t) => {
  const buffer = base64ToArrayBuffer(base64Str);
  const u8 = new Uint8Array(buffer);

  t.deepEqual([...u8], charCodeArray);
});

test('arrayBufferToString', (t) => {
  t.is(arrayBufferToString(Uint8Array.from(charCodeArray).buffer), str);
});

test('decode', (t) => {
  t.is(decode(base64Str), str);
});

function testEncoder(transformer: Transformer, t: ExecutionContext) {
  t.is(transformer.decode(), str);
  t.is(transformer.encode(), base64Str);

  const arrayBuffer = transformer.getArrayBuffer();

  if (!arrayBuffer) {
    t.fail();
  } else {
    t.deepEqual([...new Uint8Array(arrayBuffer)], charCodeArray);
  }
}

test('Transformer, 来源是原始字符串', (t) => {
  // 默认sourceType
  const transformer = new Transformer(str);
  // 指定sourceType
  const passTypeTransformer = new Transformer(str, 'string');

  testEncoder(transformer, t);
  testEncoder(passTypeTransformer, t);
});

test('Transformer, 来源是ArrayBuffer', (t) => {
  // 默认sourceType
  const transformer = new Transformer(Uint8Array.from(charCodeArray).buffer);
  // 指定sourceType
  const passTypeTransformer = new Transformer(
    Uint8Array.from(charCodeArray).buffer,
    'arrayBuffer',
  );

  testEncoder(transformer, t);
  testEncoder(passTypeTransformer, t);
});

test('Transformer, 来源是base64', (t) => {
  // 默认sourceType
  const transformer = new Transformer(base64Str, 'base64String');

  testEncoder(transformer, t);
});


src/shared/base-transformer.ts
import { SourceType } from '../type';

export abstract class BaseTransformer {
  // 缓存
  #base64String = '';
  #string = '';
  #arrayBuffer: ArrayBufferLike = new ArrayBuffer(0);

  // 数据类型
  #sourceType: SourceType;
  #source: string | ArrayBufferLike = '';

  constructor(source: string | ArrayBuffer, sourceType?: SourceType) {
    // 判断数据来源的类型
    this.#source = source;

    if (sourceType) {
      this.#sourceType = sourceType || '';
    } else {
      if (source instanceof ArrayBuffer) {
        this.#sourceType = 'arrayBuffer';
      } else {
        this.#sourceType = 'string';
      }
    }
  }

  abstract arrayBufferToBase64(buffer: ArrayBufferLike): string;
  abstract arrayBufferToString(buffer: ArrayBufferLike): string;
  abstract stringToArrayBuffer(str: string): ArrayBufferLike;
  abstract base64ToArrayBuffer(str: string): ArrayBufferLike;
  abstract _encode(str: string): string;
  abstract _decode(base64String: string): string;

  encode() {
    if (this.#base64String) return this.#base64String;

    if (this.#sourceType === 'string') {
      return (this.#base64String = this._encode(this.#source as string));
    } else if (this.#sourceType === 'arrayBuffer') {
      return (this.#base64String = this.arrayBufferToBase64(
        this.#source as ArrayBufferLike,
      ));
    } else if (this.#sourceType === 'base64String') {
      return (this.#base64String = this.#source as string);
    }

    return undefined;
  }

  decode() {
    if (this.#string) return this.#string;

    if (this.#sourceType === 'string') {
      return (this.#string = this.#source as string);
    } else if (this.#sourceType === 'arrayBuffer') {
      return (this.#string = this.arrayBufferToString(
        this.#source as ArrayBufferLike,
      ));
    } else if (this.#sourceType === 'base64String') {
      return (this.#string = this._decode(this.#source as string));
    }

    return undefined;
  }

  getArrayBuffer() {
    if (this.#arrayBuffer.byteLength) return this.#arrayBuffer;

    if (this.#sourceType === 'string') {
      return (this.#arrayBuffer = this.stringToArrayBuffer(
        this.#source as string,
      ));
    } else if (this.#sourceType === 'arrayBuffer') {
      return (this.#arrayBuffer = this.#source as ArrayBufferLike);
    } else if (this.#sourceType === 'base64String') {
      return (this.#arrayBuffer = this.base64ToArrayBuffer(
        this.#source as string,
      ));
    }

    return undefined;
  }
}


src/utf8-string/index.ts
import { BaseTransformer } from '../shared/base-transformer';

/**
 * 那字符串转换成代表着utf8的编码的arrayBuffer
 * 假设有 '一二三abc'。
 * 先用encodeUri 转成: `%E4%B8%80%E4%BA%8C%E4%B8%89abc`。
 * 然后再转成[228, 184, 128, 228, 186, 140, 228, 184, 137, 97, 98, 99]， 1字节的abc会取其charCode: 97，98，99。
 * 然后将数组转成arrayBuffer
 *
 * @param str 需要转换的字符串
 */
export function stringToArrayBuffer(str: string) {
  const encoded = encodeURIComponent(str);
  const charCodes: number[] = [];
  let i = 0;
  while (encoded[i]) {
    if (encoded[i] === '%') {
      const substr = encoded.slice(i, i + 3);
      charCodes.push(parseInt(substr.slice(-2), 16));
      i += 3;
    } else {
      charCodes.push(encoded[i].charCodeAt(0));
      i++;
    }
  }

  return Uint8Array.from(charCodes).buffer;
}

export function arrayBufferToBase64(buffer: ArrayBufferLike) {
  const u8 = new Uint8Array(buffer);
  const oneByteChars = String.fromCharCode(...u8);
  return btoa(oneByteChars);
}

export function encode(str: string) {
  return arrayBufferToBase64(stringToArrayBuffer(str));
}

export function base64ToArrayBuffer(base64: string) {
  // 获取到base64的字符串
  const binaryStr = atob(base64);
  const bytes = new Uint8Array(binaryStr.length);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = binaryStr.charCodeAt(i);
  }
  return bytes.buffer;
}

export function arrayBufferToString(buffer: ArrayBufferLike) {
  const u8 = new Uint8Array(buffer);

  const str = [...u8]
    .map((o) => {
      return `%${o.toString(16)}`;
    })
    .join('');

  return decodeURIComponent(str);
}

export function decode(base64: string) {
  return arrayBufferToString(base64ToArrayBuffer(base64));
}

export class Transformer extends BaseTransformer {
  _decode = decode;
  _encode = encode;
  arrayBufferToString = arrayBufferToString;
  base64ToArrayBuffer = base64ToArrayBuffer;
  arrayBufferToBase64 = arrayBufferToBase64;
  stringToArrayBuffer = stringToArrayBuffer;
}
